### DISCUSSION TITLE
What's the difference between error returned by Produce() and TopicPartition.Error

### DISCUSSION THREAD

**Discussion by ytlw** [confluentinc/confluent-kafka-go#1075] [Q&A]
What's the difference between these two types of errors?
1. error return by Produce https://github.com/confluentinc/confluent-kafka-go/blob/master/kafka/producer.go#L300
2. error in TopicPartition when read delivery report https://github.com/confluentinc/confluent-kafka-go/blob/v2.2.0/examples/producer_example/producer_example.go#L59

Can we handle both types of errors when handle delivery report?

**Comment by adilburaksen** [comment]
These are errors at two different stages of the message lifecycle:

**`Produce()` return error — synchronous, pre-send failure**

This error means the message was **never enqueued** for delivery. Common causes:
- `ErrQueueFull`: the producer's internal queue is at capacity (`queue.buffering.max.messages`)
- Invalid message parameters (nil topic, message too large before even contacting broker)
- Producer is in a fatal error state

If `Produce()` returns an error, you should **not** wait for a delivery report for that message — it won't come.

**`TopicPartition.Error` in the delivery report — asynchronous, post-send failure**

This error means the message **was enqueued** and sent to the broker, but delivery ultimately failed:
- Broker unavailable / leader election in progress
- Authentication/authorization failure
- Message timed out before being acked (`delivery.timeout.ms`)
- Transient network issues

The delivery channel (or `Events()` channel) receives a `*kafka.Message` with `TopicPartition.Error` set.

**Handling both:**

```go
deliveryChan := make(chan kafka.Event, 1)

// Check synchronous error first
err := p.Produce(&kafka.Message{
    TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
    Value:          []byte("payload"),
}, deliveryChan)

if err != nil {
    // Message never sent — handle immediately
    log.Printf("Produce failed: %v", err)
    return err
}

// Wait for async delivery report
e := <-deliveryChan
m := e.(*kafka.Message)
if m.TopicPartition.Error != nil {
    // Message sent but not delivered
    log.Printf("Delivery failed: %v", m.TopicPartition.Error)
    return m.TopicPartition.Error
}
// Success
```

Both errors need to be handled for robust message delivery guarantees.
